central/externalbackups/plugins/s3compatible/s3compatible.go

```diff
@@ -0,0 +1,238 @@
+	if !ok {
+		return nil, errors.New("S3 Compatible configuration required")
+	}
+	if err := validate(cfg.S3Compatible); err != nil {
+		return nil, err
+	}
+
+	opts := []func(*config.LoadOptions) error{
+		config.WithRegion(cfg.S3Compatible.GetRegion()),
+		config.WithHTTPClient(&http.Client{Transport: proxy.RoundTripper()}),
+	}
+
+	if cfg.S3Compatible.GetUseIam() {
+		opts = append(opts, config.WithCredentialsProvider(
+			credentials.NewStaticCredentialsProvider(
+				cfg.S3Compatible.GetAccessKeyId(),
+				cfg.S3Compatible.GetSecretAccessKey(), "",
+			),
+		))
+	}
+
+	ctx, cancelFn := context.WithTimeout(context.Background(), initialConfigurationMaxTimeout)
+	defer cancelFn()
+	awsConfig, err := config.LoadDefaultConfig(ctx, opts...)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to load the aws config")
+	}
+
+	var clientOpts []func(*s3.Options)
+	if cfg.S3Compatible.GetUsePathStyle() {
+		clientOpts = append(clientOpts, func(o *s3.Options) {
+			o.UsePathStyle = true
+		})
+	}
+
+	if endpoint := cfg.S3Compatible.GetEndpoint(); endpoint != "" {
+		if err := validateEndpoint(endpoint); err != nil {
+			return nil, err
+		}
+		clientOpts = append(clientOpts, func(o *s3.Options) {
+			o.BaseEndpoint = aws.String(endpoint)
+		})
+	}
+
+	client := s3.NewFromConfig(awsConfig, clientOpts...)
+	return &s3Compatible{
+		integration: integration,
+		client:      client,
+		uploader:    manager.NewUploader(client),
+	}, nil
+}
+
+func (s *s3Compatible) Backup(reader io.ReadCloser) error {
+	defer func() {
+		if err := reader.Close(); err != nil {
+			log.Errorf("closing reader: %+v", err)
+		}
+	}()
+
+	log.Info("Starting S3 Compatible Backup")
+	ctx, cancelFn := context.WithTimeout(context.Background(), backupMaxTimeout)
+	defer cancelFn()
+	formattedTime := time.Now().Format("2006-01-02T15:04:05")
+	key := fmt.Sprintf("backup_%s.zip", formattedTime)
+	formattedKey := s.prefixKey(key)
+	if _, err := s.uploader.Upload(ctx, &s3.PutObjectInput{
+		Bucket: aws.String(s.integration.GetS3Compatible().GetBucket()),
+		Key:    aws.String(formattedKey),
+		Body:   reader,
+	}); err != nil {
+		return s.createError(fmt.Sprintf("error creating backup in bucket %q with key %q",
+			s.integration.GetS3Compatible().GetBucket(), formattedKey), err)
+	}
+	log.Info("Successfully backed up to S3")
+	return s.pruneBackupsIfNecessary(ctx)
+}
+
+func (s *s3Compatible) createError(msg string, err error) error {
+	if awsErr, _ := err.(awserr.Error); awsErr != nil {
+		if awsErr.Message() != "" {
+			msg = fmt.Sprintf("%s (code: %s; message: %s)", msg, awsErr.Code(), awsErr.Message())
+		} else {
+			msg = fmt.Sprintf("%s (code: %s)", msg, awsErr.Code())
+		}
+	}
+	log.Errorf("S3 backup error: %v", err)
+	return errors.New(msg)
+}
+
+func (s *s3Compatible) prefixKey(key string) string {
+	return filepath.Join(s.integration.GetS3Compatible().GetObjectPrefix(), key)
+}
+
+func sortS3Objects(objects []s3Types.Object) {
```

Comment:
This function could be shared with `s3.go`, right?

Reply:
I could be wrong but I don't think they can be shared. `s3Types.Object` is a struct from the `aws-sdk-go-v2` and in `s3.go` we use `awsS3.Object` from `aws-sdk-go`. Since none of the structs have functions to retrieve the `LastModified` field, we need to access the field directly. This makes it impossible to pass an interface and have a unique function AFAIK.

Reply:
Ah good point. Let's keep it then.